### PR TITLE
perf(Shikimori): rewritten all code and bugs fixed 

### PR DESCRIPTION
**Pull request by shiseni**
## Description 
All code has been rewritten to switch-case. Added settings and buttons. There were also bugs on some sections that have been fixed.

Settings:
1. Private mode.
2. Show covers.
3. Show buttons.

## Acknowledgements
- [x] I read the [Presence Guidelines](https://github.com/PreMiD/Presences/blob/main/.github/CONTRIBUTING.md)
- [x] I linted the code by running `yarn format`
- [x] The PR title follows the repo's [commit conventions](https://github.com/PreMiD/Presences/blob/main/.github/COMMIT_CONVENTION.md)

## Screenshots
<details>
<summary> Proof showing the creation/modification is working as expected </summary>

image
image
image

</details>

### CODE REVIEW

File: websites/S/Shikimori/presence.ts [hunk 1]
```diff
@@ -3,152 +3,208 @@ const presence = new Presence({
...
 presence.on("UpdateData", async () => {
-	const presenceData: PresenceData = {
-			details: "Где-то на сайте",
+	const [privacy, logo, buttons] = await Promise.all([
+			presence.getSetting<boolean>("privacy"),
+			presence.getSetting<boolean>("logo"),
+			presence.getSetting<boolean>("buttons"),
+		]),
+		presenceData: PresenceData = {
 			largeImageKey: "shikimori_logo",
```
Comment (theusaf) [inline websites/S/Shikimori/presence.ts:12]: Imgur links are now preferred over Discord asset names.
```suggestion
			largeImageKey: "https://i.imgur.com/7tlc7or.png",
```

File: websites/S/Shikimori/presence.ts [hunk 2]
```diff
@@ -3,152 +3,208 @@ const presence = new Presence({
...
-		else if (document.location.pathname.includes("/completed"))
-			presenceData.state = "Просмотрено";
-		else if (document.location.pathname.includes("/on_hold"))
-			presenceData.state = "Отложено";
-		else if (document.location.pathname.includes("/dropped"))
-			presenceData.state = "Брошено";
+	function isImageExist(tags: string) {
+		return document?.querySelector<HTMLImageElement>(tags) && !privacy && logo
+			? document?.querySelector<HTMLImageElement>(tags)?.src
+			: "shikimori_logo";
```
Comment (theusaf) [inline websites/S/Shikimori/presence.ts:22]:
```suggestion
			: "https://i.imgur.com/7tlc7or.png";
```

File: websites/S/Shikimori/presence.ts [hunk 3]
```diff
@@ -3,152 +3,208 @@ const presence = new Presence({
...
+			{
+				label: "Открыть страницу",
+				url: document
+					.querySelector("meta[property='og:url']")
+					.getAttribute("content"),
+			},
+		];
+	}
+
+	switch (document.location.pathname.split("/")[1]) {
```
Comment (theusaf) [inline websites/S/Shikimori/presence.ts:36]: `document.location.pathname` is used a lot. Consider destructuring it instead.

**Review by theusaf — CHANGES_REQUESTED** [review]

File: websites/S/Shikimori/presence.ts (hunk 3 above)
Comment (shiseni) [reply websites/S/Shikimori/presence.ts:36]: > `document.location.pathname` is used a lot. Consider destructuring it instead.

Okay, I'll change it.

File: websites/S/Shikimori/presence.ts [hunk 4]
```diff
@@ -3,152 +3,206 @@ const presence = new Presence({
...
-		presenceData.state = title;
+	switch (pathname.split("/")[1]) {
+		case "":
+			presenceData.details = "На главной странице";
+			break;
+		case "animes":
+		case "mangas":
+		case "ranobe":
+			presenceData.details = `В поисках ${
+				location.pathname.split("/")[1] === "animes"
```
Comment (theusaf) [inline websites/S/Shikimori/presence.ts:45]: Missed one.
```suggestion
				pathname.split("/")[1] === "animes"
```

File: websites/S/Shikimori/presence.ts [hunk 5]
```diff
@@ -3,152 +3,206 @@ const presence = new Presence({
...
+					case "order-by":
+					case "mylist":
+					case "season":
+					case "score":
+					case "duration":
+					case "rating":
+					case "studio":
+						if (!privacy) {
+							presenceData.state =
+								document?.querySelector("header > h1")?.textContent;
```
Comment (theusaf) [inline websites/S/Shikimori/presence.ts:65]: `document` should always be defined, so there is no need to use optional chaining on it.
```suggestion
								document.querySelector("header > h1")?.textContent;
```
(same for the other instances of this.)

**Review by theusaf — CHANGES_REQUESTED** [review]

File: websites/S/Shikimori/metadata.json [hunk 6]
```diff
@@ -10,12 +10,41 @@
...
+	],
+	"settings": [
+		{
+			"id": "privacy",
+			"title": "Приватный режим",
+			"icon": "fad fa-user-secret",
+			"value": false
+		},
+		{
+			"id": "logo",
```
Comment (DarkVillager) [inline websites/S/Shikimori/metadata.json:32]: @EncryptedDev doesn't cover match other presences?

File: websites/S/Shikimori/presence.ts (hunk 5 above)
Comment (shiseni) [reply websites/S/Shikimori/presence.ts]: > `document` should always be defined, so there is no need to use optional chaining on it.
> 
> (same for the other instances of this.)

Oops, I don't know why I did it. Thanks.

File: websites/S/Shikimori/presence.ts [hunk 7]
```diff
@@ -3,152 +3,206 @@ const presence = new Presence({
...
-			presenceData.state = "Отложено";
-		else if (document.location.pathname.includes("/dropped"))
-			presenceData.state = "Брошено";
+	function isImageExist(tags: string) {
+		return document?.querySelector<HTMLImageElement>(tags) && !privacy && logo
+			? document?.querySelector<HTMLImageElement>(tags)?.src
+			: "https://i.imgur.com/7tlc7or.png";
+	}
+
+	if (buttons && !privacy) {
```
[the remark this answers is not in the document]
Comment (shiseni) [reply websites/S/Shikimori/presence.ts:25]: > Why

What's wrong? I don't understand

[the remark this answers is not in the document]
Comment (DarkVillager) [reply websites/S/Shikimori/presence.ts:25]: > > Why
> 
> What's wrong? I don't understand

Why do this multiple times?

[the remark this answers is not in the document]
Comment (shiseni) [reply websites/S/Shikimori/presence.ts:25]: > > > Why
> > 
> > 
> > What's wrong? I don't understand
> 
> Why do this multiple times?
I still don't understand what's wrong.😟
Construction is needed to display the buttons if they are enabled in the settings and the privacy mode is disabled. Am I doing it wrong?

[the remark this answers is not in the document]
Comment (DarkVillager) [reply websites/S/Shikimori/presence.ts:30]: > > > > Why
> > > 
> > > 
> > > What's wrong? I don't understand
> > 
> > 
> > Why do this multiple times?
> 
> I still don't understand what's wrong.😟 Construction is needed to display the buttons if they are enabled in the settings and the privacy mode is disabled. Am I doing it wrong?
You could just do it like [here](https://github.com/PreMiD/Presences/blob/main/websites/M/Marktplaats/presence.ts#L165)
Also no need to check for privacy cuz buttons get turned off when privacy is on.

[the remark this answers is not in the document]
Comment (shiseni) [reply websites/S/Shikimori/presence.ts]: > > > > > Why
> > > > 
> > > > 
> > > > What's wrong? I don't understand
> > > 
> > > 
> > > Why do this multiple times?
> > 
> > 
> > I still don't understand what's wrong.😟 Construction is needed to display the buttons if they are enabled in the settings and the privacy mode is disabled. Am I doing it wrong?
> 
> You could just do it like [here](https://github.com/PreMiD/Presences/blob/main/websites/M/Marktplaats/presence.ts#L165) Also no need to check for privacy cuz buttons get turned off when privacy is on.

Understood. Now it is better?
Maybe this is a bug, but my buttons remain in private mode if I remove the privacy check.

File: websites/S/Shikimori/metadata.json (hunk 6 above)
Comment (EncryptedDev) [reply websites/S/Shikimori/metadata.json:32]: Doesn't matter too much

**Review by Bas950 — APPROVED** [review]

**Pull request merged**